File: api/facebook.js
/**
 * Facebook Login: verify a user access token against this app, then load profile.
 */
const GRAPH_BASE = "https://graph.facebook.com";

function facebookAppId() {
  return (process.env.FACEBOOK_APP_ID || "").trim();
}

function facebookAppSecret() {
  return (process.env.FACEBOOK_APP_SECRET || "").trim();
}

function facebookConfigured() {
  return Boolean(facebookAppId() && facebookAppSecret());
}

function facebookPublicConfig() {
  const enabled = facebookConfigured();
  return {
    facebook_enabled: enabled,
    facebook_app_id: enabled ? facebookAppId() : null,
  };
}

/**
 * @param {string} accessToken
 * @param {{ fetch?: typeof fetch }} [deps]
 * @returns {Promise<{ ok: true, info: { id: string, email: string, name: string } } | { ok: false, error: string, status: number }>}
 */
async function verifyFacebookAccessToken(accessToken, deps = {}) {
  if (!facebookConfigured()) {
    return { ok: false, error: "Facebook sign-in is not configured.", status: 503 };
  }
  const token = typeof accessToken === "string" ? accessToken.trim() : "";
  if (!token) {
    return { ok: false, error: "Facebook access token is required.", status: 400 };
  }
  const doFetch = deps.fetch || fetch;
  const appId = facebookAppId();
  const appToken = `${appId}|${facebookAppSecret()}`;
  try {
    const debugUrl = `${GRAPH_BASE}/debug_token?input_token=${encodeURIComponent(token)}&access_token=${encodeURIComponent(appToken)}`;
    const debugResp = await doFetch(debugUrl);
    const debugBody = await debugResp.json();
    const data = debugBody.data || {};
    if (!data.is_valid || String(data.app_id || "") !== appId) {
      return { ok: false, error: "Facebook sign-in could not be verified.", status: 400 };
    }
    const userId = String(data.user_id || "");
    if (!userId) {
      return { ok: false, error: "Facebook sign-in could not be verified.", status: 400 };
    }

    const meUrl = `${GRAPH_BASE}/me?fields=id,name,email&access_token=${encodeURIComponent(token)}`;
    const meResp = await doFetch(meUrl);
    const me = await meResp.json();
    const id = String(me.id || userId);
    const email = String(me.email || "").trim().toLowerCase();
    if (!email) {
      return { ok: false, error: "Facebook did not provide an email.", status: 400 };
    }
    return { ok: true, info: { id, email, name: me.name || "" } };
  } catch {
    return { ok: false, error: "Facebook sign-in could not be verified.", status: 400 };
  }
}

function uniqueUsername(db, email, name) {
  const local = (email || "").split("@", 1)[0];
  let raw = local || name || "user";
  let cleaned = String(raw).replace(/[^A-Za-z0-9._]/g, "").slice(0, 24);
  if (!cleaned) cleaned = "user";
  if (/^\d/.test(cleaned)) cleaned = `u${cleaned}`.slice(0, 24);
  let candidate = cleaned;
  let n = 2;
  while (db.prepare("SELECT id FROM users WHERE username = ? COLLATE NOCASE").get(candidate)) {
    const suffix = String(n);
    candidate = `${cleaned.slice(0, Math.max(1, 24 - suffix.length))}${suffix}`;
    n += 1;
    if (n > 10_000) throw new Error("Could not allocate a username.");
  }
  return candidate;
}

/**
 * Find or create a USER from a verified Facebook profile.
 * @returns {{ user: object, created: boolean }}
 */
function userFromFacebook(db, info) {
  const fbId = String(info.id);
  const email = String(info.email || "").trim().toLowerCase();
  const existing = db.prepare("SELECT * FROM users WHERE facebook_id = ?").get(fbId);
  if (existing) {
    if (email && !existing.email) {
      db.prepare("UPDATE users SET email = ? WHERE id = ?").run(email, existing.id);
      existing.email = email;
    }
    return { user: existing, created: false };
  }
  if (email) {
    const byEmail = db.prepare("SELECT * FROM users WHERE email = ? COLLATE NOCASE").get(email);
    if (byEmail) {
      if (byEmail.facebook_id && byEmail.facebook_id !== fbId) {
        throw Object.assign(new Error("This Facebook account cannot be linked."), { status: 400 });
      }
      db.prepare("UPDATE users SET facebook_id = ? WHERE id = ?").run(fbId, byEmail.id);
      byEmail.facebook_id = fbId;
      return { user: byEmail, created: false };
    }
  }
  const username = uniqueUsername(db, email, info.name);
  const infoRow = db
    .prepare(
      `INSERT INTO users (username, email, password_hash, role, avatar_seed, facebook_id)
       VALUES (?, ?, '', 'USER', ?, ?)`,
    )
    .run(username, email, username, fbId);
  const user = db.prepare("SELECT * FROM users WHERE id = ?").get(infoRow.lastInsertRowid);
  return { user, created: true };
}

module.exports = {
  facebookPublicConfig,
  facebookConfigured,
  verifyFacebookAccessToken,
  userFromFacebook,
  uniqueUsername,
};

File: backend/apps/accounts/facebook.py
"""Facebook Login: verify a user access token and provision a local USER."""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request

from django.conf import settings
from django.contrib.auth import get_user_model
from rest_framework.exceptions import APIException, ValidationError

from .google import unique_username
from .models import Role

User = get_user_model()

GRAPH_BASE = "https://graph.facebook.com"


class FacebookAuthError(Exception):
    def __init__(self, detail: str, status: int = 400):
        super().__init__(detail)
        self.detail = detail
        self.status = status


def facebook_app_id() -> str:
    return (getattr(settings, "FACEBOOK_APP_ID", "") or "").strip()


def facebook_app_secret() -> str:
    return (getattr(settings, "FACEBOOK_APP_SECRET", "") or "").strip()


def facebook_configured() -> bool:
    return bool(facebook_app_id() and facebook_app_secret())


def _get_json(url: str) -> dict:
    req = urllib.request.Request(url, method="GET")
    try:
        with urllib.request.urlopen(req, timeout=8) as resp:
            return json.loads(resp.read().decode())
    except (urllib.error.URLError, TimeoutError, json.JSONDecodeError, OSError) as exc:
        raise FacebookAuthError("Facebook sign-in could not be verified.") from exc


def verify_facebook_access_token(access_token: str) -> dict:
    """Confirm the token belongs to this app and return id/name/email."""
    app_id = facebook_app_id()
    secret = facebook_app_secret()
    if not app_id or not secret:
        raise FacebookAuthError("Facebook sign-in is not configured.", status=503)
    token = (access_token or "").strip()
    if not token:
        raise FacebookAuthError("Facebook access token is required.")

    app_token = f"{app_id}|{secret}"
    debug_qs = urllib.parse.urlencode({"input_token": token, "access_token": app_token})
    debug = _get_json(f"{GRAPH_BASE}/debug_token?{debug_qs}")
    data = debug.get("data") or {}
    if not data.get("is_valid"):
        raise FacebookAuthError("Facebook sign-in could not be verified.")
    if str(data.get("app_id") or "") != app_id:
        raise FacebookAuthError("Facebook sign-in could not be verified.")
    user_id = str(data.get("user_id") or "")
    if not user_id:
        raise FacebookAuthError("Facebook sign-in could not be verified.")

    me_qs = urllib.parse.urlencode({"fields": "id,name,email", "access_token": token})
    me = _get_json(f"{GRAPH_BASE}/me?{me_qs}")
    fb_id = str(me.get("id") or user_id)
    if not fb_id:
        raise FacebookAuthError("Facebook sign-in could not be verified.")
    email = str(me.get("email") or "").strip().lower()
    if not email:
        raise FacebookAuthError("Facebook did not provide an email.")
    return {"id": fb_id, "email": email, "name": me.get("name") or ""}


def user_from_facebook(info: dict):
    """Find or create a USER from a verified Facebook profile. Returns (user, created)."""
    fb_id = str(info["id"])
    email = str(info.get("email") or "").strip().lower()

    existing = User.objects.filter(facebook_id=fb_id).first()
    if existing:
        if email and not existing.email:
            existing.email = email
            existing.save(update_fields=["email"])
        return existing, False

    if email:
        by_email = User.objects.filter(email__iexact=email).first()
        if by_email:
            if by_email.facebook_id and by_email.facebook_id != fb_id:
                raise FacebookAuthError("This Facebook account cannot be linked.")
            by_email.facebook_id = fb_id
            update = ["facebook_id"]
            if email and not by_email.email:
                by_email.email = email
                update.append("email")
            by_email.save(update_fields=update)
            return by_email, False

    user = User(
        username=unique_username(email, info.get("name")),
        email=email,
        role=Role.USER,
        facebook_id=fb_id,
    )
    user.set_unusable_password()
    user.save()
    return user, True


class FacebookSignInUnavailable(APIException):
    status_code = 503
    default_detail = "Facebook sign-in is not configured."
    default_code = "facebook_unavailable"


def raise_facebook_as_api(exc: FacebookAuthError):
    if exc.status >= 500:
        err = FacebookSignInUnavailable()
        err.detail = exc.detail
        raise err
    raise ValidationError({"detail": exc.detail})
